[SSH] Added key size selection

axmanalad opened the pull request:
This implementation allows to select the key size of the generated key ranging from 2048-4096 bits, which is using a Spinner to increment by 1024 bits. (DSA is limited to 3072 bits)

Refer to https://github.com/eclipse-platform/eclipse.platform/issues/1464

HannesWell commented on team/bundles/org.eclipse.jsch.ui/src/org/eclipse/jsch/internal/ui/messages.properties:
```diff
@@ -89,6 +89,7 @@ CVSSSH2PreferencePage_144=Key Exchange &Methods
 CVSSSH2PreferencePage_145=MA&C Methods
 CVSSSH2PreferencePage_146=&SSH Agent
 CVSSSH2PreferencePage_147=Select preferred SSH Agent
+CVSSSH2PreferencePage_148=Key Size to Generate:
```
Let's keep it simple here:
```suggestion
CVSSSH2PreferencePage_148=Key size:
```

HannesWell commented on team/bundles/org.eclipse.jsch.ui/src/org/eclipse/jsch/internal/ui/preference/PreferencePage.java:
```diff
@@ -101,6 +108,7 @@ public class PreferencePage extends org.eclipse.jface.preference.PreferencePage
 	private Button ssh2HomeBrowse;
 	Button keyGenerateDSA;
 	Button keyGenerateRSA;
+	private Spinner keyGenerateSize;
```
Better name it `keySizeValue` or just `keySize`, but then you have to rename the existing local variable with the same name to `keySizeValue` or alike or you just inline `keySize.getSelection()`
```suggestion
	private Spinner keySizeValue;
```

HannesWell commented on team/bundles/org.eclipse.jsch.ui/src/org/eclipse/jsch/internal/ui/preference/PreferencePage.java:
```diff
@@ -305,6 +313,19 @@ private Control createKeyManagementPage(Composite parent){
+		keyGenerateSize.setValues(MIN_KEY_SIZE, MIN_KEY_SIZE, MAX_KEY_SIZE, MAX_KEY_SIZE_DIGIT_COUNT,
+				KEY_SIZE_INCREMENT, KEY_SIZE_INCREMENT);
+
+		keyGenerateSize.addKeyListener(new KeyAdapter() {
+			@Override
+			public void keyPressed(KeyEvent e) {
+				e.doit = false;
+			}
+		});
```
This can be simplified using `KeyListener.keyPressedAdapter()`
```suggestion
		keySizeValue.addKeyListener(KeyListener.keyPressedAdapter(e -> e.doit = false));
```

HannesWell commented on team/bundles/org.eclipse.jsch.ui/src/org/eclipse/jsch/internal/ui/preference/PreferencePage.java:
```diff
@@ -485,6 +506,8 @@ public void widgetSelected(SelectionEvent e){
 					if(e.widget==keyGenerateDSA){
 						type=KeyPair.DSA;
 						_type=IConstants.DSA;
+						if (keyGenerateSize.getSelection()>DSA_KEY_SIZE)
+							keyGenerateSize.setSelection(DSA_KEY_SIZE);
```
Please always use braces:
```suggestion
						if (keyGenerateSize.getSelection()>DSA_KEY_SIZE) {
							keyGenerateSize.setSelection(DSA_KEY_SIZE);
						}
```

HannesWell reviewed (changes requested):
Thank you for this follow-up. Nice work, it works well.
I just have a few remarks below.

In general what do you think about setting the initial value to the recommended size of 4kBi for RSA and 3kBi for DSA and just allow lowering the value?
Are higher values possible for RSA too? If yes, I think we should also make that possible.

Additionally please make sure that you have formatted all new lines according to the formatter settings (unfortunately the existing code does not adhere to them). E.g. there should be a space before and after a equal-sign.

axmanalad commented:
> In general what do you think about setting the initial value to the recommended size of 4kBi for RSA and 3kBi for DSA and just allow lowering the value?

I think this is a good idea as it initially sets it to the more secured key size than 2048 bits for developers to acknowledge. I changed the initial value to 4096 bits.

> Additionally please make sure that you have formatted all new lines according to the formatter settings (unfortunately the existing code does not adhere to them). E.g. there should be a space before and after a equal-sign.

I have fixed the formatting. Thank you for clarification.

> Are higher values possible for RSA too? If yes, I think we should also make that possible.

Higher values for RSA are possible. The current RSA algorithm does not seem to have a limit for the max key size it can generate. However, the main drawbacks with generating larger RSA key lengths include [CPU overhead and performance issues](https://www.javamex.com/tutorials/cryptography/rsa_key_length.shtml). As you generate a larger key length, CPU runtime increases dramatically. A high-end computer could easily generate larger keys, but I would not recommend it for low-end computers.

Because of the overhead and performance, it is best to recommend using an [ECC encryption](https://ww1.microchip.com/downloads/en/DeviceDoc/00003442A.pdf) for keys larger than 4096 bits since it provides an equivalent level of encryption strength as RSA with smaller key sizes and faster performance. For example, an ECC key size of 512 bits is equivalent to an RSA key size of 15360 bits.

I changed the max RSA key size you can generate to 15360 bits as it is equivalent to 256 security bits (really strong in security encryption); generally, higher than 15360 bits is not recommended for the sake of runtime.

HannesWell commented on team/bundles/org.eclipse.jsch.ui/src/org/eclipse/jsch/internal/ui/preference/PreferencePage.java:
```diff
@@ -89,8 +91,13 @@ public class PreferencePage extends org.eclipse.jface.preference.PreferencePage
 
 	private static final String SSH2_PREFERENCE_PAGE_CONTEXT="org.eclipse.jsch.ui.ssh2_preference_page_context"; //$NON-NLS-1$
-	private static final int RSA_KEY_SIZE = 4096;
+	private static final int RSA_KEY_SIZE = 15360;
 	private static final int DSA_KEY_SIZE = 3072;
```
Let's add a `MAX` segment in these names to reflect that these are the maximum values:
```suggestion
	private static final int RSA_MAX_KEY_SIZE = 15360;
	private static final int DSA_MAX_KEY_SIZE = 3072;
```

HannesWell commented on team/bundles/org.eclipse.jsch.ui/src/org/eclipse/jsch/internal/ui/preference/PreferencePage.java:
```diff
@@ -89,8 +91,13 @@ public class PreferencePage extends org.eclipse.jface.preference.PreferencePage
 	private static final int DSA_KEY_SIZE = 3072;
+	private static final int INITIAL_KEY_SIZE = 4096;
+	private static final int MIN_KEY_SIZE = 2048;
+	private static final int MAX_KEY_SIZE = RSA_KEY_SIZE;
+	private static final int MAX_KEY_SIZE_DIGIT_COUNT = 0;
```
These constants are not needed.
```suggestion
```

HannesWell commented on team/bundles/org.eclipse.jsch.ui/src/org/eclipse/jsch/internal/ui/preference/PreferencePage.java:
```diff
@@ -305,6 +313,13 @@ private Control createKeyManagementPage(Composite parent){
+		final Label keySizeValueLabel = new Label(group, SWT.NONE);
+		keySizeValueLabel.setText(Messages.CVSSSH2PreferencePage_148);
+		keySizeValue = new Spinner(group, SWT.BORDER);
+		keySizeValue.setValues(INITIAL_KEY_SIZE, MIN_KEY_SIZE, MAX_KEY_SIZE, MAX_KEY_SIZE_DIGIT_COUNT,
+				KEY_SIZE_INCREMENT, KEY_SIZE_INCREMENT);
```
This can be structured a bit more to prevent an over-long line and two constants:
```suggestion
		int maxKeySize = Math.max(DSA_MAX_KEY_SIZE, RSA_MAX_KEY_SIZE);
		keySizeValue.setValues(INITIAL_KEY_SIZE, MIN_KEY_SIZE, maxKeySize, 0, KEY_SIZE_INCREMENT, KEY_SIZE_INCREMENT);
```

HannesWell commented on team/bundles/org.eclipse.jsch.ui/src/org/eclipse/jsch/internal/ui/preference/PreferencePage.java:
```diff
@@ -485,6 +500,9 @@ public void widgetSelected(SelectionEvent e){
 					if(e.widget==keyGenerateDSA){
 						type=KeyPair.DSA;
 						_type=IConstants.DSA;
+						if (keySizeValue.getSelection() > DSA_KEY_SIZE) {
+							keySizeValue.setSelection(DSA_KEY_SIZE);
```
Adapt to new constant names
```suggestion
						if (keySizeValue.getSelection() > DSA_MAX_KEY_SIZE) {
							keySizeValue.setSelection(DSA_MAX_KEY_SIZE);
```

HannesWell reviewed (commented):
Thanks for the update and sorry for the deplayed reply.
I just have a few minor style remarks below.
Technically this is fine. 👍🏽 

But we are currently in the quiet-period before the December release and have to await the start of the next development cycle, which should start in one to two weeks, before this can be submitted. But in the meantime we can make it ready.

> Because of the overhead and performance, it is best to recommend using an [ECC encryption](https://ww1.microchip.com/downloads/en/DeviceDoc/00003442A.pdf) for keys larger than 4096 bits since it provides an equivalent level of encryption strength as RSA with smaller key sizes and faster performance. For example, an ECC key size of 512 bits is equivalent to an RSA key size of 15360 bits. I am not sure how implementing ECC would work but it is only a thought I think would improve performance in generating keys.

It would be great if we could also generated more modern keys and if you are interested in looking into it, you are more than welcome.
I'm not sure, but I fear that the currently used library is not capable of these newer algorithms.
In fact we actually would like to replace JSCH with something that's maintained better and more up-to-date, but havn't done it yet due to a lack of known simple replacements, see
- https://github.com/eclipse-platform/eclipse.platform/issues/958

But if you are interested in that please don't hesitate to look into it. :)

axmanalad commented:
No worries, it is understandable. I reformatted the variables as requested.

> It would be great if we could also generated more modern keys and if you are interested in looking into it, you are more than welcome. I'm not sure, but I fear that the currently used library is not capable of these newer algorithms. In fact we actually would like to replace JSCH with something that's maintained better and more up-to-date, but havn't done it yet due to a lack of known simple replacements, see
> 
> * [Replace or remove org.eclipse.jsch plugins? #958](https://github.com/eclipse-platform/eclipse.platform/issues/958)
> 
> But if you are interested in that please don't hesitate to look into it. :)

While looking in the #958, I believe replacing the JSch library is a major change to be made. I lack the knowledge to knowing how to make these replacements however. In the meantime, I looked deeper into the codebase to find that there exists an ECC algorithm in the JSch library known as **ECDSA** before the library became outdated. This is a type of ECC algorithm that does exactly what I previously stated and is found within "*com.jcraft.jsch.KeyPairECDSA*". Maybe after this PR, I can make a third contribution to add the feature of generating ECDSA keys in a new PR, which is why I added a "TODO" comment on Line 509.

HannesWell commented:
Sorry, I somehow missed the notification about your latest update, will look into it as soon as possible.

> > It would be great if we could also generated more modern keys and if you are interested in looking into it, you are more than welcome. I'm not sure, but I fear that the currently used library is not capable of these newer algorithms. In fact we actually would like to replace JSCH with something that's maintained better and more up-to-date, but havn't done it yet due to a lack of known simple replacements, see
> > 
> > * [Replace or remove org.eclipse.jsch plugins? #958](https://github.com/eclipse-platform/eclipse.platform/issues/958)
> > 
> > But if you are interested in that please don't hesitate to look into it. :)
> 
> While looking in the #958, I believe replacing the JSch library is a major change to be made. I lack the knowledge to knowing how to make these replacements however. In the meantime, I looked deeper into the codebase to find that there exists an ECC algorithm in the JSch library known as **ECDSA** before the library became outdated. This is a type of ECC algorithm that does exactly what I previously stated and is found within "_com.jcraft.jsch.KeyPairECDSA_". Maybe after this PR, I can make a third contribution to add the feature of generating ECDSA keys in a new PR, which is why I added a "TODO" comment on Line 509.

That would be very nice. But we should be cautious and should check how well that implementation works. Having a flawed implementation of a per-se save algorithm would pretend false security, which would be really bad.

HannesWell reviewed (approved):
This looks perfectly fine now.
Thanks for the update and sorry for the delayed reply. I lost track of this during the hectic rush before Christmas.
